onnxruntime/contrib_ops/cpu/quantization/qlinear_softmax.cc

```diff
@@ -0,0 +1,344 @@
+                                  lookup_table.data(), Y_scale, Y_zp, thread_pool);
+  }
+  return status;
+}
+
+// opset-13 and above
+Status QLinearSoftmax::ComputeImplOpset13(OpKernelContext* context,
+                                          const Tensor& input, Tensor& output,
+                                          gsl::span<const uint32_t> lookup_table,
+                                          concurrency::ThreadPool* thread_pool) const {
+  const auto& X_shape = input.Shape();
+  size_t rank = X_shape.NumDimensions();
+
+  bool is_transpose_required = (size_t(axis_) != (rank - 1));
+  Tensor transposed_input;
+  Tensor intermediate_output;  // output that the softmax implementation will write into while using transposed input
+  std::vector<size_t> permutation(rank);
+
+  if (is_transpose_required) {
+    AllocatorPtr alloc;
+    ORT_RETURN_IF_ERROR(context->GetTempSpaceAllocator(&alloc));
+    std::iota(std::begin(permutation), std::end(permutation), 0);
+
+    // swap the innermost dim with the dim corresponding to axis
+    permutation[axis_] = rank - 1;
+    permutation[rank - 1] = axis_;
+    std::vector<int64_t> transposed_input_dims(rank);
+    std::transform(permutation.cbegin(), permutation.cend(),
+                   transposed_input_dims.begin(), [&X_shape](size_t e) { return X_shape[e]; });
+
+    // Allocate a temporary tensor to hold transposed input
+    transposed_input = Tensor(input.DataType(), TensorShape(transposed_input_dims), alloc);
+    // Perform the transpose
+    ORT_RETURN_IF_ERROR(TransposeBase::DoTranspose(permutation, input, transposed_input));
+    // Allocate memory for the intermediate output
+    intermediate_output = Tensor(output.DataType(), TensorShape(transposed_input_dims), alloc);
+  }
+
+  common::Status status;
+
+  const auto& input_tensor = is_transpose_required ? transposed_input : input;
+  auto& output_tensor = is_transpose_required ? intermediate_output : output;
+
+  ORT_RETURN_IF_ERROR(ComputeInternal(context, input_tensor, output_tensor, lookup_table, int(rank - 1), thread_pool));
+
+  if (is_transpose_required) {
+    // Perform the transpose to get the axes back to the original ordering
+    status = (TransposeBase::DoTranspose(permutation, intermediate_output, output));
+  }
+  return status;
+}
+
+#define REGISTER_QLINEAR_LOOKUPTABLE_TYPED_KERNEL(op_name, version, data_type, KERNEL_CLASS) \
+  ONNX_CPU_OPERATOR_TYPED_MS_KERNEL(                                                         \
+      op_name, version, data_type,                                                           \
+      KernelDefBuilder()                                                                     \
+          .TypeConstraint("T", DataTypeImpl::GetTensorType<data_type>()),                    \
+      KERNEL_CLASS);
+
+REGISTER_QLINEAR_LOOKUPTABLE_TYPED_KERNEL(QLinearSoftmax, 1, uint8_t, QLinearSoftmax);
+REGISTER_QLINEAR_LOOKUPTABLE_TYPED_KERNEL(QLinearSoftmax, 1, int8_t, QLinearSoftmax);
```

Comment:
You can register 2 types with one register: ONNX_CPU_OPERATOR_MS_KERNEL and TypeConstraint to include both int8_t and uint8_t

Reply:
Sure.  Thanks